nixos/modules/services/misc/gitlab.nix

```diff
@@ -598,21 +591,33 @@ in {
 
         # JSON is a subset of YAML
         ln -fs ${pkgs.writeText "gitlab.yml" (builtins.toJSON gitlabConfig)} ${cfg.statePath}/config/gitlab.yml
-        ln -fs ${pkgs.writeText "database.yml" databaseYml} ${cfg.statePath}/config/database.yml
-        ln -fs ${pkgs.writeText "secrets.yml" secretsYml} ${cfg.statePath}/config/secrets.yml
+        cp -f  ${pkgs.writeText "database.yml" databaseYml} ${cfg.statePath}/config/database.yml
+        cp -f  ${pkgs.writeText "secrets.yml" secretsYml} ${cfg.statePath}/config/secrets.yml
         ln -fs ${pkgs.writeText "unicorn.rb" unicornConfig} ${cfg.statePath}/config/unicorn.rb
 
         chown -R ${cfg.user}:${cfg.group} ${cfg.statePath}/
         chmod -R ug+rwX,o-rwx+X ${cfg.statePath}/
 
+        DBPASSWORD=$(cat ${cfg.databasePassword})
+        SECRETKEY=$(cat ${cfg.secrets.secret})
+        OTPKEY=$(cat ${cfg.secrets.otp})
+        DBKEY=$(cat ${cfg.secrets.db})
+        JWSKEY=$(${pkgs.jq}/bin/jq -Rs . ${cfg.secrets.jws} | sed -e 's,\\n,\\\\n,g')
+
+        sed -e "s,#dbpasswordplaceholder#,$DBPASSWORD,g" -i ${cfg.statePath}/config/database.yml
+        sed -e "s,#secretkeyplaceholder#,$SECRETKEY,g"   -i ${cfg.statePath}/config/secrets.yml
+        sed -e "s,#otpkeyplaceholder#,$OTPKEY,g"         -i ${cfg.statePath}/config/secrets.yml
+        sed -e "s,#dbkeyplaceholder#,$DBKEY,g"           -i ${cfg.statePath}/config/secrets.yml
+        sed -e "s,#jwskeyplaceholder#,$JWSKEY,g"         -i ${cfg.statePath}/config/secrets.yml
+
         # Install the shell required to push repositories
         ln -fs ${pkgs.writeText "config.yml" gitlabShellYml} "$GITLAB_SHELL_CONFIG_PATH"
         ln -fs ${cfg.packages.gitlab-shell}/hooks "$GITLAB_SHELL_HOOKS_PATH"
         ${cfg.packages.gitlab-shell}/bin/install
 
         if [ "${cfg.databaseHost}" = "127.0.0.1" ]; then
           if ! test -e "${cfg.statePath}/db-created"; then
-            ${pkgs.sudo}/bin/sudo -u ${pgSuperUser} psql postgres -c "CREATE ROLE ${cfg.databaseUsername} WITH LOGIN NOCREATEDB NOCREATEROLE ENCRYPTED PASSWORD '${cfg.databasePassword}'"
+            ${pkgs.sudo}/bin/sudo -u ${pgSuperUser} psql postgres -c "CREATE ROLE ${cfg.databaseUsername} WITH LOGIN NOCREATEDB NOCREATEROLE ENCRYPTED PASSWORD '$DBPASSWORD'"
```

Comment:
This would break with a password containing a quote, right? Similarly to before, I'd suggest `${DBPASSWORD//'/\\'/}`… Although I'm guessing that would still go wrong in other ways 😢 (I wouldn't be surprised if postgres interprets other escape sequences too.) Oh, the joys of substituting text into other languages.

Reply:
Also, shouldn't bash refrain from variable substitutions in single quotes anyway?

Reply:
Uh, yes. I must have been fooled by the peer authentication.

Reply:
Oh, no, this is fine.

    $ foo=test; echo "This is fine: '$foo'"
    This is fine: 'test'

Reply:
Hm, weird. _Without_ the outer quotes:

```
foo=test; echo This is fine: '$foo'
This is fine: $foo
```

Reply:
It's not that weird. The single quotes' quotyness is quoted by the outer double quotes, so they are just literal single quote characters. :)

Reply:
:-)

Reply:
(Section 3.1.2.3 of the bash manual)[https://www.gnu.org/software/bash/manual/bash.html#Double-Quotes] on how double quotes work elaborates on this. :)

Reply:
`'$DBPASSWORD'` should not be passed on the `sudo` command line that creates the database, as it will show up in the output of `ps` if someone happens to run it at the right time. It would be better to write the `postgres` command into a file readable only by `pgSuperUser` and then `cat` that to `stdin`.

Reply:
Good point. Although there's no need for writing a temporary file, we can just use bash:

```
${pkgs.sudo}/bin/sudo -u ${pgSuperUser} psql postgres <<-EOF
CREATE ROLE ${cfg.databaseUsername} WITH LOGIN NOCREATEDB NOCREATEROLE ENCRYPTED PASSWORD '$DBPASSWORD'
EOF
```

Reply:
Would assigning the query to a variable and echo-piping it to psql leak the password? I feel sort of bad relying on tab-indentation in the heredoc.

Reply:
Ah, I guess that has the same problem that @dhess pointed out.

Reply:
Yes, because it would appear in `ps`, just for echo rather than psql. You could also remove the `-`, since psql won't care about the leading whitespace AFAIK.

Reply:
Another alternative is `psql postgres <<< "CREATE ROLE ${cfg.databaseUsername} WITH LOGIN NOCREATEDB NOCREATEROLE ENCRYPTED PASSWORD '$DBPASSWORD'"`

Reply:
How about 4c9744b ?

Reply:
LGTM!